Documentation/devel/debugging.md

```diff
@@ -0,0 +1,195 @@
+```
+
+
+Luckily, rkt includes a helper script that injects a static busybox binary and makes entering stage1 easier.
+It can be found in `scripts/debug/stage1_install_busybox.sh`.
+It takes a pod UUID as a parameter and it outputs the right nsenter command:
+
+```bash
+$ rkt list --full
+UUID					APP	IMAGE NAME			IMAGE ID		STATE	CREATED					STARTED			NETWORKS
+86e6df38-0762-4261-af21-d2b265555179	busybox	kinvolk.io/aci/busybox:1.24	sha512-140375b2a2bd	running	2017-12-19 14:49:00.376 +0100 CET	2017-12-19 14:49:00.466 +0100 CET	default:ip4=172.16.28.27
+$ ./scripts/debug/stage1_install_busybox.sh 86e6df38-0762-4261-af21-d2b265555179
+Busybox installed. Use the following command to enter pod's stage1:
+sudo nsenter -m -u -i -p -t 8505
+-bash-4.3# cat rkt/env/busybox
+SHELL=/bin/sh
+USER=root
+LOGNAME=root
+HOME=/root
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+AC_APP_NAME=busybox
+```
+
+By default, the script creates link for common busybox programs, but if you need to run more commands you can call busybox accordingly:
+
+```bash
+-bash-4.3# busybox head -n2 /lib/systemd/system/busybox.service
+[Unit]
+After=systemd-journald.service
+```
+
+If you need custom tools you can add them by copying them to `/proc/$PID/root/bin`.
+It's easier if they're static binaries to avoid complications with dynamic libraries.
+
+Run `busybox` without arguments to see a list of available commands.
+
+## Debugging images
+
+As mentioned, `rkt enter` has limitations when the image to debug doesn't ship with tools to debug it or even a shell.
+One possible way to debug those images is to use the ACI dependencies mechanism to add a custom tools image to your container.
+
+Let's use [actool](https://github.com/appc/spec/tree/v0.8.11/actool) to patch the manifest.
+
+```bash
+$ rkt fetch debug-image.aci
+sha512-847f2ca60e473121311fac62d014a4f1
+$ actool cat-manifest --pretty-print image-to-patch.aci > manifest.json
+```
+
+We'll now add the dependency to the manifest, right before the "app" section:
+
+```json
+    ],
+    "dependencies": [
+        {
+            "imageID": "sha512-847f2ca60e473121311fac62d014a4f1"
+        }
+    ],
+    "app": {
+```
+
+Then we'll patch the image and run it:
+
+```bash
+$ actool patch-manifest --manifest manifest.json --replace image-to-patch.aci
+$ sudo rkt --insecure-options=image run --interactive image-to-patch.aci
+(...)
+```
+
+You'll have your favorite debugging tools available.
+
+For examples of how to build images, check out [Build container examples](../build-container-examples).
+Creating images with statically built binaries is preferred to avoid complications with shared libraries.
+
+As an alternative to adding a dependency in the ACI, you can also add a volume after a container has been started using [`machinectl bind`](https://www.freedesktop.org/software/systemd/man/machinectl.html#bind%20NAME%20PATH%20%5BPATH%5D):
+
+```bash
+$ sudo mkdir /proc/$PID/root/opt/stage2/$APP_NAME/rootfs/tools
```

Comment:
$PID of the app or of systemd?

Reply:
It was systemd, but I guess I can simplify it and use the PID of the app?

Reply:
Or thinking about it, since we need to use the `/opt/...` path with `machinectl bind` maybe I can keep it the way it was and specify that PID is the PID of systemd. Wdyt?